Fix double speaking label in angular checkbox

feerrenrut opened the pull request:
<!--
Please fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
-->

### Link to issue number:
https://bugs.chromium.org/p/chromium/issues/detail?id=995603

### Summary of the issue:
The HTML for checkboxes on https://dart-lang.github.io/angular_components/#/material_checkbox have `aria-labelledby` pointing to a sub-element of the checkbox. Normally, we announce the accessible name as we descend then read the child elements. This results in double speaking the name.

### Description of how this pull request fixes the issue:
When rendering the virtual buffer, track the ID's of the nodes. If this node has a label, check if the ID for the label node matches any sub-node then set an attribute `labelledByContents`. Then in `speech.py` don't speak the name for a `controlField` if it has this attribute.

### Testing performed:
Simplified test case:
```
data:text/html,<button>begin</button><div tabindex="0" role="checkbox" aria-labelledby="inner-label"><div style="display:inline" id="inner-label">Simulate evil cat</div></div>
```

### Known issues with pull request:

### Change log entry:

Section: New features, Changes, Bug fixes

feerrenrut committed:
Fix double speaking content when labelled by a child element

feerrenrut committed:
Rename and refactor

Improve names, comments and code structure.

feerrenrut committed:
remove unused IdMap

michaelDCurran reviewed (commented):
I don't quite get what parentIDs does? I get that outIds stores all descendant IDs, and that if the labelledID appears in the list, then you mark it as the label coming from content.
An out of the box idea though:
Since the buffer already stores nodes by ID, rather than using your own ID lists, why not just make use of what the buffer has already. So call GetControlfieldNodeByIdentifier with the label ID, once before all the possible recursions, and then once after, at the end. If a node did not exist before, but does after, than the node is a descendant. Note that getControlFieldNodeByIdentifier is very cheep as it itself just looks up the ID in a map.

feerrenrut commented:
> I don't quite get what parentIDs does?

While developing this I wasn't sure if I would need to check parent ID's on the child side, and left it in for the symmetry.

I see what you are suggesting, with `getControlFieldNodeWithIdentifier`, it's a clever approach, but I think it makes the code harder to maintain, since it implicitly depends on how that map is built. Though, looking at how that map is built (via the calls to `addControlFieldNode`), in the same place I'll build an ID tree.

feerrenrut committed:
Build a tree of Node Identifiers instead of passing parameters.

During addControlFieldNode a mapping
of parent->list[immediate children IDs] is built.

This can then be used to get all the children for a node, or all parents
for a node if necessary.

feerrenrut committed:
Use is descendant and existing node tree instead.

The node tree already has all the information necessary, so removing
the node ID map.

maniekins commented:
I noticed, the aria-labelledby is not working with inner elements after 2019.3 version and it is probably related with this changes. I tested example 2 from https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-labelledby_attribute , added some button next to header, and when we are focusing on that button, main is vocalized but without h1 label.
